_posts/2024-04-15-heapdump-oome-native.adoc

```diff
@@ -0,0 +1,258 @@
+author: galderz
+---
+
+Starting with GraalVM for JDK 21,
+native executables can run with the `-XX:+HeapDumpOnOutOfMemoryError` option to generate a heap dump when a `java.lang.OutOfMemoryError` is thrown.
+In this blog post we will explore how to use the flag, we will inspect what a native image heap dump looks like and how it compares with one produced by HotSpot.
+
+To see this flag in action, we need to manufacture a situation where a Quarkus application runs out of memory.
+One easy way to achieve this is to configure the application with a garbage collector that doesn't do any memory reclamation,
+and that's exactly what the Epsilon GC does.
+Once the Quarkus application is running with Epsilon GC,
+apply some load and within a short space of time it will run out of memory and produce a heap dump.
+
+Let’s start by producing a heap dump with a Quarkus native executable.
+A good starting point would be a Quarkus application that simply responds to an HTTP endpoint request, e.g. the
+https://github.com/quarkusio/quarkus-quickstarts/tree/main/getting-started-reactive[getting-started-reactive] Quarkus quickstart.
+Once checked out,
+build a Quarkus native executable with GraalVM for JDK 21 configuring Epsilon GC.
+For Quarkus native mode, selecting the GC needs to be done at build time:
+
+[source,bash]
+----
+$ ./mvnw package -DskipTests -Dnative -Dquarkus.native.additional-build-args=--gc=epsilon
+----
+
+While the Quarkus native executable is being produced you will be able to observe that GC is indeed configured to be EpsilonGC:
+
+[source,bash]
+----
+[INFO] --- quarkus-maven-plugin:3.9.3:build (default) @ getting-started-reactive ---
+...
+[1/8] Initializing...
+ Java version: 21.0.2+13, vendor version: GraalVM CE 21.0.2+13.1
+ Graal compiler: optimization level: 2, target machine: compatibility
+ C compiler: gcc (redhat, x86_64, 13.2.1)
+ Garbage collector: Epsilon GC (max heap size: 80% of RAM)
+----
+
+Once the build completes,
+start Quarkus with `-XX:+HeapDumpOnOutOfMemoryError -XX:+ExitOnOutOfMemoryError`.
+The latter forces the application to shutdown when an `OutOfMemoryError` occurs rather than leave the process in an indeterminate state:
+
+[source,bash]
+----
+$ getting-started-reactive-1.0.0-SNAPSHOT-runner -XX:+HeapDumpOnOutOfMemoryError -XX:+ExitOnOutOfMemoryError -Xmx64m
+__  ____  __  _____   ___  __ ____  ______
+ --/ __ \/ / / / _ | / _ \/ //_/ / / / __/
+ -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \
+--\___\_\____/_/ |_/_/|_/_/|_|\____/___/
+2024-04-12 09:58:00,855 INFO  [io.quarkus] (main) getting-started-reactive 1.0.0-SNAPSHOT native (powered by Quarkus 3.9.3) started in 0.030s. Listening on: http://0.0.0.0:8080
+2024-04-12 09:58:00,855 INFO  [io.quarkus] (main) Profile prod activated.
+2024-04-12 09:58:00,855 INFO  [io.quarkus] (main) Installed features: [cdi, rest, smallrye-context-propagation, vertx]
+----
+
+Next, apply some load to the HTTP endpoint exposed by the Quarkus application, for example using `wrk`:
```

Comment:
Perhaps a bash loop with a curl command would be better as most people won't have `wrk` installed. On the other hand that's a nice way to become aware of `wrk`.

Reply:
Good point, `wrk` is an overkill for this. Just used to quickly firing it up :)